### ISSUE TITLE
[Bug]: (not necessary) package check not work properly

### ISSUE THREAD

**Issue by ananosleep** [Mikubill/sd-webui-controlnet#2961]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

WebUI throw out `changing opencv-python version from 4.10.0.84 to 4.8.0` every time, 
although it's already satisfied `opencv-python>=4.8.0`

### Steps to reproduce the problem

Just do a simple launch.

### What should have happened?

Ignore packages which are already satisfied

### Commit where the problem happens

webui: v1.9.4-169-ga30b19dd
controlnet: 8bbbd0e

### What browsers do you use to access the UI ?

Microsoft Edge

### Command Line Arguments

```Shell
--xformers --no-gradio-queue --opt-channelslast --precision half
```

### List of enabled extensions

this extension itself only.

### Console logs

```Shell
Python 3.10.14+ (heads/3.10:48f9d3e3fa, Jun 10 2024, 16:53:46) [MSC v.1939 64 bit (AMD64)]
Version: v1.9.4-169-ga30b19dd
Commit hash: a30b19dd5536f463222e484aef2daf466b49ee85
Installing sd-webui-controlnet requirement: changing opencv-python version from 4.10.0.84 to 4.8.0
Looking in indexes: https://pypi.org/simple
Requirement already satisfied: opencv-python>=4.8.0 in d:\windows\sd-webui\stable-diffusion-webui-dev\venv2\lib\site-packages (4.10.0.84)
Requirement already satisfied: numpy>=1.21.2 in d:\windows\sd-webui\stable-diffusion-webui-dev\venv2\lib\site-packages (from opencv-python>=4.8.0) (1.26.2)
Launching Web UI with arguments: --xformers --no-gradio-queue --opt-channelslast --precision half --lora-dir E:\temp-model
ldm/sgm GroupNorm32 replaced with normal torch.nn.GroupNorm due to `--precision half`.
ControlNet preprocessor location: D:\Windows\sd-webui\A1111-extensions\sd-webui-controlnet\annotator\downloads
2024-06-21 18:34:46,433 - ControlNet - INFO - ControlNet v1.1.449
Loading weights [2a5740e96c] from D:\Windows\sd-webui\stable-diffusion-webui-dev\models\Stable-diffusion\_Für-Alice-fp8.safetensors
Creating model from config: D:\Windows\sd-webui\stable-diffusion-webui-dev\configs\v1-inference.yaml
D:\Windows\sd-webui\stable-diffusion-webui-dev\venv2\lib\site-packages\huggingface_hub\file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
2024-06-21 18:34:47,365 - ControlNet - INFO - ControlNet UI callback registered.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 29.7s (prepare environment: 11.2s, import torch: 9.1s, import gradio: 1.7s, setup paths: 2.6s, initialize shared: 0.5s, other imports: 1.2s, list SD models: 0.1s, load scripts: 1.6s, create ui: 1.3s, gradio launch: 0.2s).
Applying attention optimization: xformers... done.
Model loaded in 11.8s (load weights from disk: 0.2s, create model: 1.6s, apply weights to model: 9.3s, apply half(): 0.2s, load textual inversion embeddings: 0.2s, calculate empty prompt: 0.2s).
```

**Comment by JayNaire** [comment]
Me too...
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
Version: v1.9.4
Commit hash: feee37d75f1b168768014e4634dcb156ee649c05
Installing sd-webui-controlnet requirement: changing opencv-python version from 4.10.0.84 to 4.8.0
...

**Comment by RoyalXXX** [comment]
I also got the line **Installing sd-webui-controlnet requirement: changing opencv-python version from 4.10.0.84 to 4.8.0** after installing controlnet

**Comment by yxyh357** [comment]
> ### Is there an existing issue for this?
> 这是否存在现有问题？
> * [x]  I have searched the existing issues and checked the recent builds/commits of both this extension and the webui 我已经搜索了现有问题并检查了此扩展和 webui 的最近构建/提交
> 
> ### What happened? 发生了什么？
> WebUI throw out `changing opencv-python version from 4.10.0.84 to 4.8.0` every time,WebUI 每次都会抛出 `changing opencv-python version from 4.10.0.84 to 4.8.0` ， although it's already satisfied `opencv-python>=4.8.0` 虽然已经很满足 `opencv-python>=4.8.0`
> 
> ### Steps to reproduce the problem
> 重现问题的步骤
> Just do a simple launch.只需进行简单的启动即可。
> 
> ### What should have happened?
> 应该发生什么？
> Ignore packages which are already satisfied忽略已经满意的包
> 
> ### Commit where the problem happens
> 提交问题发生的地方
> webui: v1.9.4-169-ga30b19dd网络用户界面：v1.9.4-169-ga30b19dd controlnet: [8bbbd0e](https://github.com/Mikubill/sd-webui-controlnet/commit/8bbbd0e55ef6e5d71b09c2de2727b36e7bc825b0)  控制网： 8bbbd0e
> 
> ### What browsers do you use to access the UI ?
> 您使用什么浏览器来访问 UI？
> Microsoft Edge 微软边缘
> 
> ### Command Line Arguments 命令行参数
> ```shell
> --xformers --no-gradio-queue --opt-channelslast --precision half
> ```
> 
> ### List of enabled extensions
> 已启用的扩展列表
> this extension itself only.仅此扩展本身。
> 
> ### Console logs 控制台日志
> ```shell
> Python 3.10.14+ (heads/3.10:48f9d3e3fa, Jun 10 2024, 16:53:46) [MSC v.1939 64 bit (AMD64)]
> Version: v1.9.4-169-ga30b19dd
> Commit hash: a30b19dd5536f463222e484aef2daf466b49ee85
> Installing sd-webui-controlnet requirement: changing opencv-python version from 4.10.0.84 to 4.8.0
> Looking in indexes: https://pypi.org/simple
> Requirement already satisfied: opencv-python>=4.8.0 in d:\windows\sd-webui\stable-diffusion-webui-dev\venv2\lib\site-packages (4.10.0.84)
> Requirement already satisfied: numpy>=1.21.2 in d:\windows\sd-webui\stable-diffusion-webui-dev\venv2\lib\site-packages (from opencv-python>=4.8.0) (1.26.2)
> Launching Web UI with arguments: --xformers --no-gradio-queue --opt-channelslast --precision half --lora-dir E:\temp-model
> ldm/sgm GroupNorm32 replaced with normal torch.nn.GroupNorm due to `--precision half`.
> ControlNet preprocessor location: D:\Windows\sd-webui\A1111-extensions\sd-webui-controlnet\annotator\downloads
> 2024-06-21 18:34:46,433 - ControlNet - INFO - ControlNet v1.1.449
> Loading weights [2a5740e96c] from D:\Windows\sd-webui\stable-diffusion-webui-dev\models\Stable-diffusion\_Für-Alice-fp8.safetensors
> Creating model from config: D:\Windows\sd-webui\stable-diffusion-webui-dev\configs\v1-inference.yaml
> D:\Windows\sd-webui\stable-diffusion-webui-dev\venv2\lib\site-packages\huggingface_hub\file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
>   warnings.warn(
> 2024-06-21 18:34:47,365 - ControlNet - INFO - ControlNet UI callback registered.
> Running on local URL:  http://127.0.0.1:7860
> 
> To create a public link, set `share=True` in `launch()`.
> Startup time: 29.7s (prepare environment: 11.2s, import torch: 9.1s, import gradio: 1.7s, setup paths: 2.6s, initialize shared: 0.5s, other imports: 1.2s, list SD models: 0.1s, load scripts: 1.6s, create ui: 1.3s, gradio launch: 0.2s).
> Applying attention optimization: xformers... done.
> Model loaded in 11.8s (load weights from disk: 0.2s, create model: 1.6s, apply weights to model: 9.3s, apply half(): 0.2s, load textual inversion embeddings: 0.2s, calculate empty prompt: 0.2s).
> ```
> 
> ### Additional information 附加信息
> _No response 没有反应_

我找到了原因:extensions\sd-webui-controlnet\install.py下的comparable_version函数返回了一个字符串数组并进行比较,这使得"10">"8"不成立,我将函数修改为def comparable_version(version: str) -> Tuple:
    return tuple(int(i) for i in version.split("."))后起作用了

**Comment by yxyh357** [comment]
"I have found the cause of this issue. In install.py, line 16, the function comparable_version return a tuple[str] for comparison. Due to '10' being less than '8' in lexicographical order, it resulted in an erroneous version rollback."

**Comment by JayNaire** [comment]
Yay! Thanks for that @yxyh357 :+1: 
To fix : 
edit _stable-diffusion-webui/extensions/sd-webui-controlnet/install.py_  and change line 16:
```
def comparable_version(version: str) -> Tuple:
    #return tuple(version.split(".")) existing line 16 commented out
    return tuple(map(int, version.split(".")))  # fixed line

```

**Comment by silvertuanzi** [comment]
seems to be fixed at 56cec5b2958edf3b1807b7e7b2b1b5186dbd2f81
